File: src/2025-02-23-deepclaude-try/java/exercises/practice/state-of-tic-tac-toe/src/main/java/StateOfTicTacToe.java
class StateOfTicTacToe {
    public GameState determineState(String[] board) {
        // Count X's and O's to verify turn order
        int xCount = 0;
        int oCount = 0;
        for (String row : board) {
            for (char c : row.toCharArray()) {
                if (c == 'X') xCount++;
                if (c == 'O') oCount++;
            }
        }

        // Verify valid turn order (X starts, so X count should be equal to or one more than O count)
        if (oCount > xCount || xCount > oCount + 1) {
            throw new IllegalArgumentException("Invalid turn order");
        }

        // Check for wins
        boolean xWin = hasWon('X', board);
        boolean oWin = hasWon('O', board);

        // Check invalid states
        if ((xWin && oWin) || // Both players can't win
            (xWin && xCount <= oCount) || // If X wins, they must have more marks
            (oWin && xCount > oCount)) { // If O wins, counts must be equal
            throw new IllegalArgumentException("Invalid board state");
        }

        // Return game state
        if (xWin || oWin) {
            return GameState.WIN;
        }

        // Check if board is full (draw)
        boolean isFull = true;
        for (String row : board) {
            if (row.contains(" ")) {
                isFull = false;
                break;
            }
        }

        return isFull ? GameState.DRAW : GameState.ONGOING;
    }

    private boolean hasWon(char player, String[] board) {
        // Check rows
        for (String row : board) {
            if (row.equals("" + player + player + player)) {
                return true;
            }
        }

        // Check columns
        for (int col = 0; col < 3; col++) {
            if (board[0].charAt(col) == player &&
                board[1].charAt(col) == player &&
                board[2].charAt(col) == player) {
                return true;
            }
        }

        // Check diagonals
        if (board[0].charAt(0) == player &&
            board[1].charAt(1) == player &&
            board[2].charAt(2) == player) {
            return true;
        }

        if (board[0].charAt(2) == player &&
            board[1].charAt(1) == player &&
            board[2].charAt(0) == player) {
            return true;
        }

        return false;
    }
}

File: src/2025-02-23-deepclaude-try/javascript/exercises/practice/state-of-tic-tac-toe/state-of-tic-tac-toe.js
export const gamestate = (board) => {
  // Parse board into 2D array
  const rows = board.split('\n').filter(row => !row.includes('___') && row.trim() !== '');
  const grid = rows.map(row => 
    row.split('|').map(cell => cell.trim())
  );

  // Count X's and O's
  let xCount = 0, oCount = 0;
  for (let i = 0; i < grid.length; i++) {
    for (let j = 0; j < grid[i].length; j++) {
      if (grid[i][j] === 'X') xCount++;
      if (grid[i][j] === 'O') oCount++;
    }
  }

  // Validate turn order
  if (oCount > xCount) {
    throw new Error('Wrong turn order: O started');
  }
  if (xCount > oCount + 1) {
    throw new Error('Wrong turn order: X went twice');
  }

  // Check for wins
  const checkWin = (player) => {
    // Check rows
    for (let i = 0; i < 3; i++) {
      if (grid[i][0] === player && grid[i][1] === player && grid[i][2] === player) {
        return true;
      }
    }
    // Check columns 
    for (let i = 0; i < 3; i++) {
      if (grid[0][i] === player && grid[1][i] === player && grid[2][i] === player) {
        return true;
      }
    }
    // Check diagonals
    if (grid[0][0] === player && grid[1][1] === player && grid[2][2] === player) {
      return true;
    }
    if (grid[0][2] === player && grid[1][1] === player && grid[2][0] === player) {
      return true;
    }
    return false;
  };

  const xWin = checkWin('X');
  const oWin = checkWin('O');

  // Validate game didn't continue after win
  if ((xWin || oWin) && xCount + oCount > 5) {
    throw new Error('Impossible board: game should have ended after the game was won');
  }

  // Return game state
  if (xWin || oWin) {
    return 'win';
  }
  if (xCount + oCount === 9) {
    return 'draw';
  }
  return 'ongoing';
};
